File: Runtime/Plugins/iOS/Source/AffiseNativeModule.swift
import AffiseInternal

@objc 
public class AffiseNativeModule : NativeCallHandler {

    private static let KeyDelay = "AffiseDeepLinkStartDelay"
    private static let DefaultDelay = 3
    
    @objc 
    public static let shared = AffiseNativeModule()
        
    private var apiWrapper: AffiseApiWrapper? = nil
       
    public func setCallback(callback: @escaping (String, String) -> Void) {
        apiWrapper?.setCallback { (apiName: String, data: [String: Any?]) in
            DispatchQueue.main.async {
                callback(apiName, data.toArray().jsonString())
            }
        }
    }
    
    @objc(startAffiseWithOptions:)
    public func startAffise(_ launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
        let app = UIApplication.shared
        apiWrapper = AffiseApiWrapper(app, launchOptions: launchOptions)
        apiWrapper?.unity()
    }
    
    @objc(handleDeeplink:)
    public func handleDeeplink(_ url: URL?) {
        guard let link = url?.absoluteString else { return }
        apiWrapper?.handleDeeplink(link)
    }

    override func apiCall(_ api: AffiseApiMethod, data: [String: Any?], result: inout AffiseResult) {
        apiWrapper?.call(api, map: data, result: result)
    }

    func getDelay() -> Int {
        let delayKey = Bundle.main.object(forInfoDictionaryKey: AffiseNativeModule.KeyDelay) as? NSNumber
        let delay = delayKey?.intValue ?? AffiseNativeModule.DefaultDelay
        return delay
    }
}


extension AffiseNativeModule : ModuleAppDelegate {

    @objc
    public func applicationWillFinishLaunchingWithOptions(_ notification: Notification) {
        let options = notification.userInfo as? [UIApplication.LaunchOptionsKey : Any]
        startAffise(options)
        
        guard let url = options?[UIApplication.LaunchOptionsKey.url] as? URL else { return }
        DispatchQueue.global().asyncAfter(deadline: .now() + TimeInterval(getDelay())) {
            self.handleDeeplink(url)
        }
    }
    
    @objc
    public func onOpenURL(_ notification: Notification) {
        let options = notification.userInfo as? [String:Any?]
        let url: URL? = options?["url"] as? URL
        
        handleDeeplink(url)
    }
}

File: Runtime/Plugins/unity.androidlib/src/main/java/com/affise/attribution/unity/AffiseNativeModule.java
package com.affise.attribution.unity;

import android.app.Activity;

import com.affise.attribution.internal.AffiseApiMethod;
import com.affise.attribution.internal.AffiseApiWrapper;
import com.affise.attribution.internal.callback.InternalResult;
import com.affise.attribution.internal.callback.OnAffiseCallback;

import java.util.Map;

public class AffiseNativeModule extends NativeCallHandler {

    private final AffiseApiWrapper apiWrapper;

    public AffiseNativeModule(Activity activity, OnAffiseCallback callback) {
        apiWrapper = new AffiseApiWrapper(activity.getApplication());
        apiWrapper.setActivity(activity);
        apiWrapper.unity();
        apiWrapper.setCallback(callback);
    }

    @Override
    public void apiCall(AffiseApiMethod api, Map<String, ?> data, InternalResult result) {
        if (apiWrapper == null) return;
        apiWrapper.call(api, data, result);
    }
    
    public void nativeHandleDeeplink(String url) {
        if (apiWrapper == null) return;
        apiWrapper.handleDeeplink(url);
    }
}
